### ISSUE TITLE
Add views to kingfisher-process pipeline

### ISSUE THREAD

**Issue by duncandewhurst** [open-contracting/kingfisher-views#31]
Sometimes helpdesk analysts get requests for analysis on very short notice (1-2 days) and waiting for views to run overnight can mean we need to do analysis on the 'raw' data in kingfisher-process, which is much more time consuming.

Instead of running views as a nightly job which takes a long time to run, could we instead add creating the views for a collection to the 'standard' kingfisher-process pipeline? i.e. once all other transforms have completed on a collection, run views on that collection only.

We could have a configuration option for this, so we have it switched on for the helpdesk instance but off by default etc.

cc @pindec @mrshll1001

**Comment by duncandewhurst** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Add views to kingfisher-process pipeline](https://trello.com/c/qvQfGSre/101-add-views-to-kingfisher-process-pipeline)

**Comment by pindec** [comment]
Faster views related to the process pipeline and/or views triggered on demand for individual collections would be very helpful.

**Comment by duncandewhurst** [comment]
This will be even more useful now that I am working in a different timezone (I just local loaded some data on Tuesday morning here, but will likely need to wait until Thursday for the views to be updated).

**Comment by odscjames** [comment]
>  views triggered on demand for individual collections would be very helpful.

We would like to do this as a first solution, as this can be started and done reasonably quickly. 

> could we instead add creating the views for a collection to the 'standard' kingfisher-process pipeline?

To do this we would have to set up a process communication between process and views and while that's doable, it would take more time to discuss a good solution for this.

**Comment by odscjames** [comment]
Tho would want to check with @kindly that it's not a problem to trigger a single collection out of the normal schedule, and there aren't any aggregate stats across all collections.

**Comment by jpmckinney** [comment]
I figure for the first solution (which I agree is worthwhile to do first) we can add an option to the `refresh-views` command to accept one or more collection IDs?

For the second option, we were discussing having Process publish a pub/sub message to a 'collection done' channel. In that case, Views can subscribe to that channel.

**Comment by odscjames** [comment]
I've had a chat with @kindly and it turns out it is more complex due to the schema stuff. In brief, the current system destroys and rebuilds everything in one go and there is no provision for adding extra data in later. This was done due to the plan at the time to make it easy for analysts with SQL writing skills to get involved in adding functionality to the system. 

@kindly has however already been thinking about solutions to this and he will chat to @pindec @duncandewhurst and report back.

**Comment by pindec** [comment]
From discussions with @kindly - one approach is to generate a separate schema and views for each collection id - e.g. `select * from views_66.release_summary`.

Where we need to compare data within collection ids (e.g. schema upgrade work), then we need a mechanism to create a separate schema and views for a particular set of collection_ids e.g. `select * from views_set_1.release_summary`.

**Comment by odscjames** [comment]
https://github.com/open-contracting/kingfisher-views/pull/47 has been done and deployed. Minimal Salt work has been done to make sure it will deploy fine. More salt work is needed to make it actually working. This will be done very soon.

This
* Makes a way that analysts can by hand create a view for any collection at any time and run it instantly 
* Makes a way more read only users can be added, these users will have access to everything (Unfortuently the way we added such users before was just broken by this, so it was essential I did this at the same time)

There is more to discuss here, but first I will complete the salt work to make this fully work, fully test it and then introduce it to the analysts.

**Comment by odscjames** [comment]
Extra salt work and basic testing on live server done. https://github.com/open-contracting/kingfisher-views/pull/50 makes calling this much easier. Waiting to speak to analysts.

**Comment by odscjames** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Make Kingfisher Views Run Automatically when Process has finished with a data set](https://trello.com/c/8TFYiLAm/162-make-kingfisher-views-run-automatically-when-process-has-finished-with-a-data-set)

**Comment by odscjames** [comment]
ANALYSTS:

> could we instead add creating the views for a collection to the 'standard' kingfisher-process pipeline? i.e. once all other transforms have completed on a collection, run views on that collection only.

Can you clarify fully the user requirement here? When a scraper is started to add data to Kingfisher, exactly which collections should be made into views automatically?

**Comment by jpmckinney** [comment]
Also, do we always want views for every crawl, or should we make it so that analysts 'opt-in' to having views automatically generated after that crawl. (Analysts can always manually trigger the views creation if they change their mind.)

**Comment by robredpath** [comment]
@odscjames I'm not sure that anyone will actually have seen your question - could you edit to directly notify/@mention  the relevant people?

**Comment by duncandewhurst** [comment]
> Can you clarify fully the user requirement here? When a scraper is started to add data to Kingfisher, exactly which collections should be made into views automatically?

For release spiders:

* Generate views for the `upgrade-1-0-to-1-1` and `compile-releases` transformed collections

For record spiders:

* Generate views for the `upgrade-1-0-to-1-1` transformed collection

> Also, do we always want views for every crawl, or should we make it so that analysts 'opt-in' to having views automatically generated after a given crawl? (Analysts can always manually trigger the views creation if they change their mind.)

I'm happy with opting-in, as long as the example commands for running a spider and running a local-load in the kingfisher documentation are updated to have the flag turned on (as I am in the habit of copy-pasting these)

cc @pindec @mrshll1001 @yolile @romifz @camilamila for your thoughts

**Comment by romifz** [comment]
@duncandewhurst why do we need the views from `upgrade-1-0-to-1-1` to be generated automatically? I usually create collections for new publishers, so I don't even have the need for the upgrade transformation. Is this something we need often?

**Comment by duncandewhurst** [comment]
> @duncandewhurst why do we need the views from `upgrade-1-0-to-1-1` to be generated automatically? I usually create collections for new publishers, so I don't even have the need for the upgrade transformation. Is this something we need often?

We apply the upgrade transformation to all collections so that we always have data in the same format and don't need to write different queries to deal with 1.0 and 1.1.

I think some publishers are still on 1.0, so it just makes it simpler to always use the transformed collection, rather than remembering which are which, even if most publishers are on 1.1.

**Comment by jpmckinney** [comment]
Ideally (and likely in the future) we would have the spiders tell Process whether an upgrade is needed, to avoid unnecessary processing. In that case, the behavior would be to create views for the upgraded collection (if there is one) and to otherwise create views for the original collection.

**Comment by odscjames** [comment]
We discussed this at the call today:

* Maybe we just create a view for all collections as they come in (raw and compiled) as disk space is not excessive?  (If coupled with a way to delete semi-automatically?) Need to check numbers.
* Maybe we have flags that analysts can set when starting a scrape to say they would like a view (either raw or compiled or both) at the end. This bypasses trying to work out a standard requirement at this time.

It was noted however that this is on hold as the rewrite of process means we can't think about any technical plans for this at the moment.

**Comment by jpmckinney** [comment]
I prefer flags (the rewrite will preserve the standard pipeline, but there will be new options to effectively opt-out of transforms, as these aren't always needed, e.g. if you just want to run checks).

How will Views know that Process wants it to create a particular view? In the rewrite, for the compile transform for example, a message is sent to RabbitMQ with a payload of `{"collection_id": 123}` and a routing key of "compile-releases"; workers then consume such messages from a "compile-releases" queue. To coordinate with Views, we can:

* Use RabbitMQ, and route similar messages to a "views" queue (if the collection is so configured), for Views to exclusively consume. Ideally, Views would have an always-on worker to receive any messages (the pattern for this will have been established in Process).
* Use some other technology – though that means Process will be adding a dependency just for communicating with Views.

For coordinating with the Data Quality Tool: since it already uses RabbitMQ internally, we just need to publish a message to its instance.

**Comment by jpmckinney** [comment]
Closing this issue, as the main user need was for Views to be available quickly, which they are now. (Running time is an hour or less according to logs, and will be minutes once #120 is merged.)

**Comment by jpmckinney** [comment]
Follow-up issue for next version of Kingfisher Process: https://github.com/open-contracting/kingfisher-process/issues/296
